---
title: "20210812_scifate_DEG"
author: "Kathleen Abadie"
date: "8/12/2021"
output: html_document
---

```{r}
# detach(package:monocle)
suppressMessages(library(cowplot))
suppressMessages(library(tidyverse))
suppressMessages(library(monocle3))
suppressMessages(library(plotly))
library(irlba)

library(EnhancedVolcano)
library(ComplexHeatmap)
library(circlize)


library(dplyr) 
library(reshape2)
library(RColorBrewer)
library(tibble)
library(readxl)
library(Matrix.utils)
library(ComplexHeatmap)
library(circlize)

suppressMessages(library(umap))
suppressMessages(library(ggplot2))

theme_set(theme_cowplot(font_size=7, line_size = 0.5))
options(repr.plot.width=3, repr.plot.height=3)
rm(list = ls())

```


# load data, set output folder 
```{r}
#source("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/scripts_from_jun/20200612_gene_module_analysis_functions.R")

# location to store output 
output_folder <- "/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202101_output"
output_folder_new <- "/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202108_output"

# cds_all is the cds data for full mRNA
# cds_new is the cds data for new mRNA only
load("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/2020Nova_Seq/sci-fate_T_analysis_Wei/data/cds_T_cell_cycle_monocle3.RDS") # this contains cds_new and cds_all objects --> monocle3

# df_TF_net is the TF-gene linkage data identified
df_gene_TF_all_full <- readRDS('/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/2020Nova_Seq/GS_cluster_output/Wei_latest/Link_full_mgi/df_gene_TF_all.RDS')

df_gene_TF_all_new <- readRDS('/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/2020Nova_Seq/GS_cluster_output/Wei_latest/Link_new_mgi/df_gene_TF_all.RDS')

duprows <- which(!is.na(match(df_gene_TF_all_new$TF_link,df_gene_TF_all_full$TF_link)))
df_TF_net <- rbind(df_gene_TF_all_full,df_gene_TF_all_new[-duprows,] )
#df_TF_net = read_csv("~/Projects/processed_data/181103_sciMM/pdf/Main_A549_2/Gene_prediction/df_TF_net.csv")

# df_cell with cluster info and pseudotime
df_cell <- readRDS(file.path(output_folder, "df_cell_5k_Tcell_7clust.RDS"))
# individual trajectory df_cell with pseudotime
df_cell_E1 <- readRDS("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202101_output/df_cell_E1.RDS")
df_cell_E2 <- readRDS("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202101_output/df_cell_E2.RDS")
df_cell_MP <- readRDS("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202101_output/df_cell_MP.RDS")

# Additional info
TF_Tcell <- c('JUNB', 'MYC', 'TBX21', 'IRF4', 'EGR1', 'NFATC1', 'BMYC', 'MXI1', 'STAT5A', 'NFAT5', 'ELK3', 'EOMES', 'REL', 'BHLHE40', 'STAT3', 'RUNX2', 'FOXO3', 'MXD4', 'BCL11B', 'CUX1', 'GTF2I', 'FOXO1', 'FLI1', 'STAT1', 'CHD2', 'ZEB1', 'FOXN3', 'TCF7', 'LEF1', 'ELF1', 'MYB', 'IKZF1', 'TCF12')

# load scifate functions
source("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/scripts_from_jun/2021_scifate_analysis_functions.R")

```

#Plotting settings - colors and annotations
```{r}
# ordered
color_cluster_new = c("cyan4", "#08519C",
                      "#969696", "#cc6541", "#A50F15",
                      "#3F007D", "#ab62c0")

GoI_volcano <- c('ID3', 'MYC', 'XCL1', 'BCL2', 'FOXO1', 'CCL3', 'CCL4', 'CXCR4', 'TOX', 'CXC4', 'KLF6', 'CCR7', 'GZMB','IL2RA','NFATC1','CCl3','CCL4','NR4A3','NR4A2','MYC','IFNG','TBX21','ZEB2','IL12RB2','BHLHE40','BACH2','SLAMF6','LEF1','TCF7','SELL','MYB', 'ZEB2', 'BCL2', 'PECAM1', 'BCL6', 'AFF3')

volc_theme = theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), plot.background = element_blank(), axis.line = element_line(colour = "black"), text=element_text(size=16) )
```

#Filter
```{r}
# filter cds by UMI
cds_all <- cds_all[,colData(cds_all)$mRNA_count > 5000] 
cds_new <- cds_new[,colData(cds_new)$mRNA_count > 5000] 

# add newly_syn_UMIs col to cds_all 
colData(cds_all)$newly_syn_UMIs <- colData(cds_new)$newly_syn_UMIs

# make sure df_cell and cds_all and cds_new have same cells
cds_all <- cds_all[ , which(colData(cds_all)$sample %in% df_cell$sample)] 
cds_new <- cds_new[ , which(colData(cds_new)$sample %in% df_cell$sample)] 

# estimate size factors
cds_new = estimate_size_factors(cds_new)
cds_all = estimate_size_factors(cds_all)

# detect genes
cds_new <- detect_genes(cds_new)
cds_all <- detect_genes(cds_all)
```


# Perform different differential gene expression tests
1. For heatmap,all timepoints, all clusters
2. volcano: day 2, cluster 1a and 5 (MPA and E1A)
3. volcano: all timepoints, cluster 3/4
4. volcano: all timepoints, cluster 1b/4
5. volcano: all timepoints, cluster 1b/3


```{r}
# Function to perform DEG between two clusters, optionally subsetting by timepoint
pairwise_DEG <- function(cds_subset, clust_select = NULL, time_select_startswith = NULL, clust_relevel = NULL){
  
  # Subset to only include select clusters
  if (!is.null(clust_select)) {
  cds_subset <- cds_subset[, colData(cds_subset)$tcell_cluster %in% clust_select]}
  
  # Subset to only include select timepoints
  if (!is.null(time_select_startswith)) {
    cds_subset <- cds_subset[,startsWith(colData(cds_subset)$RT_group, time_select_startswith )] }
  
  # Relevel cluster factors so that DEG analysis compares all clusters to first
  if (!is.null(clust_relevel)) {
  colData(cds_subset)$tcell_cluster <- factor(colData(cds_subset)$tcell_cluster, levels = clust_relevel)}
  
  # Fit 
  gene_fits  = fit_models(cds_subset, model_formula_str = "~tcell_cluster")
  fit_coefs= coefficient_table(gene_fits)
  cluster_terms <- fit_coefs%>% filter(term != "(Intercept)")
  cluster_terms <-cluster_terms %>% filter (q_value < 0.05) %>%
    dplyr::select(gene_short_name, term, q_value, estimate)
  
  return(cluster_terms)
}

# Run DEG function on all subsets of interest

# Remove genes expressed in < 50 cells (for all DEG tests) 
cds_subset <- cds_all[which(rowData(cds_all)$num_cells_expressed > 50), ]

# add cluster info to cds
colData(cds_subset)$tcell_cluster <- df_cell$Tcell_5k_aligned_7clust

# all timepoints, all clusters
# cluster_terms_all_clust_all_time <- pairwise_DEG(cds_subset, clust_select=NULL, time_select_startswith=NULL, clust_relevel=c('2a', '5', '4', '2b', '3', '1a', '1b'))

# day 2 clusters
# cluster_terms_c1a_c5_48h <- pairwise_DEG(cds_subset, clust_select=c('1a','5'), time_select_startswith='4') 

# all timepoints, pairwise comparisons
# cluster_terms_c3_c4 <- pairwise_DEG(cds_subset, clust_select=c('3','4'))
# cluster_terms_c1b_c4 <- pairwise_DEG(cds_subset, clust_select=c('1b','4'))
# cluster_terms_c1b_c3 <- pairwise_DEG(cds_subset, clust_select=c('1b','3'))

# save individual cluster terms
# saveRDS(cluster_terms_all_clust_all_time, file = file.path(output_folder_new, 'cluster_terms_all_clust_all_time.Rda'))
# saveRDS(cluster_terms_c1a_c5_48h, file = file.path(output_folder_new, 'cluster_terms_c1a_c5_48h.Rda'))
# saveRDS(cluster_terms_c3_c4, file = file.path(output_folder_new, 'cluster_terms_c3_c4.Rda'))
# saveRDS(cluster_terms_c1b_c4, file = file.path(output_folder_new, 'cluster_terms_c1b_c4.Rda'))
# saveRDS(cluster_terms_c1b_c3, file = file.path(output_folder_new, 'cluster_terms_c1b_c3.Rda'))

```

# Read previously saved cluster terms
```{r}
# day 2 clusters
cluster_terms_c1a_c5_48h <- readRDS(file.path(output_folder_new, 'cluster_terms_c1a_c5_48h.Rda'))

# all timepoints, pairwise comparisons
cluster_terms_c3_c4 <- readRDS(file.path(output_folder_new, 'cluster_terms_c3_c4.Rda'))
cluster_terms_c1b_c4 <- readRDS(file = file.path(output_folder_new, 'cluster_terms_c1b_c4.Rda'))
cluster_terms_c1b_c3 <- readRDS(file.path(output_folder_new, 'cluster_terms_c1b_c3.Rda'))

# all clusters all time
cluster_terms_all_clust_all_time <- readRDS(file.path(output_folder_new, 'cluster_terms_all_clust_all_time.Rda'))

```


# Save DEG to table for supplemental tables
```{r}
# df key connecting cluster numbers to identity
num <- c('1a', '1b', '2a', '2b', '3', '4', '5')
clust_ID <- c('MP(A)', 'MP(B)', 'U', 'E2(A)', 'E2(B)', 'E1(B)', 'E1(A)')

# save cluster terms to csv
add_cluster_label <- function(cluster_terms, num, clust_ID){

  # load 
  cluster_terms_towrite <- cluster_terms
  
  # add cluster ID column next to term 
  cluster_terms_towrite <- cluster_terms_towrite %>% mutate(
    tcell_cluster_ID = case_when(grepl(num[1], term) ~ clust_ID[1],
                                grepl(num[2], term) ~ clust_ID[2],
                                grepl(num[3], term) ~ clust_ID[3],
                                grepl(num[4], term) ~ clust_ID[4],
                                grepl(num[5], term) ~ clust_ID[5],
                                grepl(num[6], term) ~ clust_ID[6],
                                grepl(num[7], term) ~ clust_ID[7]
  ), .after = 'term')
  
  return(cluster_terms_towrite)
}

# all clusters all time
cluster_terms_all_clust_all_time <- readRDS(file.path(output_folder_new, 'cluster_terms_all_clust_all_time.Rda'))
cluster_terms_all_clust_all_time_towrite <- add_cluster_label(cluster_terms_all_clust_all_time, num, clust_ID)
write_csv(cluster_terms_all_clust_all_time_towrite, file.path(output_folder_new, 'supplementary_tables', 'cluster_terms_all_clust_all_time.csv'))

# c1a_c5_48h - MP(A)_E1(A)_48h
cluster_terms_c1a_c5_48h <- readRDS(file.path(output_folder_new,'cluster_terms_c1a_c5_48h.Rda'))
cluster_terms_c1a_c5_48h_towrite <- add_cluster_label(cluster_terms_c1a_c5_48h, num, clust_ID)
write_csv(cluster_terms_c1a_c5_48h_towrite, file.path(output_folder_new, 'supplementary_tables', 'cluster_terms_MP(A)_E1(A)_48h.csv'))

# c3_c4 - E2(B)_E1(B)
cluster_terms_c3_c4 <- readRDS(file.path(output_folder_new, 'cluster_terms_c3_c4.Rda'))
cluster_terms_c3_c4_towrite <- add_cluster_label(cluster_terms_c3_c4, num, clust_ID)
write_csv(cluster_terms_c3_c4_towrite, file.path(output_folder_new, 'supplementary_tables', 'cluster_terms_E2(B)_E1(B).csv'))

# c1b_c4 - MP(B)_E1(B)
cluster_terms_c1b_c4 <- readRDS(file.path(output_folder_new, 'cluster_terms_c1b_c4.Rda'))
cluster_terms_c1b_c4_towrite <- add_cluster_label(cluster_terms_c1b_c4, num, clust_ID)
write_csv(cluster_terms_c1b_c4_towrite, file.path(output_folder_new, 'supplementary_tables', 'cluster_terms_MP(B)_E1(B).csv'))

# c1b_c3 - MP(BP)_E2(B)
cluster_terms_c1b_c3 <- readRDS(file.path(output_folder_new, 'cluster_terms_c1b_c3.Rda'))
cluster_terms_c1b_c3_towrite <- add_cluster_label(cluster_terms_c1b_c3, num, clust_ID)
write_csv(cluster_terms_c1b_c3_towrite, file.path(output_folder_new, 'supplementary_tables', 'cluster_terms_MP(BP)_E2(B).csv'))  


```

# Prepare aggregated matrix 
```{r}

prep_agg_mat <- function(cds_subset, df_cell, cluster_terms, clust_select = NULL,  time_select_startswith = NULL, topn = NULL, GoI = NULL){
  
  # Subset to only include select clusters
  if (!is.null(clust_select)) {
    cds_subset <- cds_subset[, colData(cds_subset)$tcell_cluster %in% clust_select]}
  
  # Subset to only include select timepoints
  if (!is.null(time_select_startswith)) {
    cds_subset <- cds_subset[,startsWith(colData(cds_subset)$RT_group, time_select_startswith )] }
  
  # Subset to only include top n DEG -- for heatmap
  if (!is.null(topn)) {
    cluster_terms_dist <- cluster_terms %>% arrange(q_value) %>% distinct(gene_short_name, .keep_all = TRUE)
    cluster_terms_top_n <- cluster_terms_dist %>% top_n(-topn, q_value)
    cluster_terms_GoI <- cluster_terms_dist %>% filter(gene_short_name %in% GoI) %>% filter(q_value < 1e-10)
    cluster_terms_top_n_GoI <- rbind(cluster_terms_top_n, cluster_terms_GoI) %>% distinct(gene_short_name, .keep_all = TRUE)
    cluster_terms <- cluster_terms_top_n_GoI }
  
  # from fit_models result 
  cds_subset_DE <- cds_subset[rowData(cds_subset)$gene_short_name %in% cluster_terms$gene_short_name,]
  # for rare case when a gene short name is duplicated, remove it
  cds_subset_DE <- cds_subset_DE[!duplicated(rowData(cds_subset_DE)$gene_short_name), ] 
  
  df_cell_subset_DE <- df_cell[which(df_cell$sample %in% colData(cds_subset_DE)$sample), ]
  cell_group_df <- tibble::tibble(cell=row.names(colData(cds_subset_DE)), 
                                  cell_group=colData(cds_subset_DE)$tcell_cluster)
     
  # aggregate expression                                                          
  agg_mat <- aggregate_gene_expression(cds_subset_DE, gene_group_df = NULL, cell_group_df, norm_method = "size_only")
  row.names(agg_mat) <- stringr::str_c(rowData(cds_subset_DE)$gene_short_name)
  colnames(agg_mat) <- stringr::str_c("Cluster ", colnames(agg_mat))
  
  return(agg_mat)
  }
  
# Prepare agg_mat for all analyses

# all timepoints, all clusters (for heatmap)
agg_mat_all_clust_all_time <- prep_agg_mat(cds_subset, df_cell, cluster_terms_all_clust_all_time, clust_select=NULL, time_select_startswith=NULL,topn = 400, GoI <- c('IFNG'))

# day 2 clusters
# agg_mat_c1a_c5_48h <- prep_agg_mat(cds_subset, df_cell, cluster_terms_c1a_c5_48h, clust_select=c('1a','5'), time_select_startswith='4') 
# 
# # all timepoints, pairwise comparisons
# agg_mat_c3_c4 <- prep_agg_mat(cds_subset, df_cell, cluster_terms_c3_c4, clust_select=c('3','4'))
# agg_mat_c1b_c4 <- prep_agg_mat(cds_subset, df_cell, cluster_terms_c1b_c4, clust_select=c('1b','4'))
# agg_mat_c1b_c3 <- prep_agg_mat(cds_subset, df_cell, cluster_terms_c1b_c3, clust_select=c('1b','3'))

```


# Make volcano plots of each pairwise DEG comparison using respective agg_mat generated above
```{r}

make_volcano_v2 <- function(agg_mat, cluster_terms, genes, lit_info_df = NULL, label_color_pair_1 = NULL, label_color_pair_2 = NULL){

  # make df from matrix and adjust naming
  agg_mat_df <- as.data.frame(as.matrix(agg_mat))
  agg_mat_df$gene_short_name <- str_to_title(tolower(row.names(agg_mat)))
  rownames(agg_mat_df) <- str_to_title(tolower(row.names(agg_mat_df)))
agg_mat_df <- agg_mat_df %>% 
  mutate(FC= .[[1]] /.[[2]]) 

  # identify FC numerator and denominator for subsequent labeling and coloring
  FC_top = colnames(agg_mat_df)[1]
  FC_bot = colnames(agg_mat_df)[2]
  if (FC_top %in% label_color_pair_1){
    pos_FC_col <- label_color_pair_1[3]
    pos_FC_lab <- label_color_pair_1[2]
    neg_FC_col <- label_color_pair_2[3]
    neg_FC_lab <- label_color_pair_2[2]
  } else {
    pos_FC_col <- label_color_pair_2[3]
    pos_FC_lab <- label_color_pair_2[2]
    neg_FC_col <- label_color_pair_1[3]
    neg_FC_lab <- label_color_pair_1[2]
  }
  
  # add column with q value from DEG test
  cluster_terms$gene_short_name <- str_to_title(tolower(cluster_terms$gene_short_name))
  agg_mat_df$q_value <- cluster_terms[match(agg_mat_df$gene_short_name,cluster_terms$gene_short_name), ]$q_value
  agg_mat_df <- agg_mat_df %>% mutate(log2FC = log2(FC))
  agg_mat_df <- agg_mat_df %>% mutate(`-log10(q_value)` = -log10(q_value))

  # remove any rows with logFC Inf
  agg_mat_df <- agg_mat_df[which(is.finite(agg_mat_df$log2FC)==TRUE), ]
  
  # Color by umap cluster
  keyvals <- ifelse(
    agg_mat_df$log2FC <= -.5, neg_FC_col,
      ifelse(agg_mat_df$log2FC >= .5, pos_FC_col,
        'grey'))
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals == neg_FC_col] <- neg_FC_lab
  names(keyvals)[keyvals == pos_FC_col] <- pos_FC_lab
  names(keyvals)[keyvals == 'grey'] <- 'NS'
  
  # label genes of interest 
  agg_mat_df_FC <- agg_mat_df[abs(agg_mat_df$log2FC) >= .5,  ] # subset with FC > 0.5
  genes_to_label = genes[genes %in% agg_mat_df[abs(agg_mat_df$log2FC) >= .5,  ]$gene_short_name]
  # and label top q val and fc
  genes_to_label <- c(genes_to_label, 
                      agg_mat_df[which.max(abs(agg_mat_df$log2FC)), ]$gene_short_name,
                      agg_mat_df_FC[which.min(abs(agg_mat_df_FC$q_value)), ]$gene_short_name)

  # axis limits
  xmax <- max(agg_mat_df$log2FC)
  xmin <- min(agg_mat_df$log2FC)
  ymax <- max(agg_mat_df[is.finite(agg_mat_df$`-log10(q_value)`), ]$`-log10(q_value)`) # accounting for possible Inf
  ymin <- min(agg_mat_df$`-log10(q_value)`)

  
  # put labels in italics
  lab_italics <- paste0("italic('", agg_mat_df$gene_short_name, "')")
  selectLab_italics = paste0(
    "italic('",
    genes_to_label,
    "')")
  
    # Run enhanced volcano
  EnhancedVolcano(agg_mat_df,
                   # lab = agg_mat_df$gene_short_name,
                  lab = lab_italics,
                   x = 'log2FC',
                   y='q_value',
                  xlim = c(xmin, xmax),
                  ylim = c(ymin,ymax),
                  FCcutoff = .5,
                  pCutoff = 0.05,
                  pointSize = 1,
                  title = NULL,
                  subtitle = NULL, 
                  xlab = "",
                  ylab = "",
                  # selectLab = genes_to_label,
                  selectLab = selectLab_italics,
                  parseLabels = TRUE,
                  labSize = 6,
                  colCustom = keyvals,
                  drawConnectors = TRUE,
                  max.overlaps = Inf,
                  widthConnectors = 0.3,
                  typeConnectors = 'open',
                  colConnectors = "grey30",
                  colAlpha = 10/10, 
                  cutoffLineType = "blank",
                  legendPosition = 'none', 
                  caption = ""
                   )
  
  }
  
# make volcano plot 
genes = str_to_title(tolower(GoI_volcano))

# Day2, c1a vs c5
volc_c1a_c5_48h = make_volcano_v2(agg_mat_c1a_c5_48h, cluster_terms_c1a_c5_48h, genes, lit_info_df = DEG_immgen_mem_eff_1000, label_color_pair_1 = c("Cluster 1a", "Memory Precursor A", "cyan4"), label_color_pair_2 = c("Cluster 5", "Effector 1A","#ab62c0")) + volc_theme

# all days, c3 vs c4
volc_c3_c4 = make_volcano_v2(agg_mat_c3_c4, cluster_terms_c3_c4, genes, lit_info_df = DEG_immgen_mem_eff_1000, label_color_pair_1 = c("Cluster 3", "E2(B)", "#A50F15"), label_color_pair_2 = c("Cluster 4", "E1(B)","#3F007D")) + volc_theme

# all days, c1b vs c4
volc_c1b_c4 = make_volcano_v2(agg_mat_c1b_c4, cluster_terms_c1b_c4, genes, lit_info_df = DEG_immgen_mem_eff_1000, label_color_pair_1 = c("Cluster 1b", "Memory Precursor B", "#08519C"), label_color_pair_2 = c("Cluster 4", "Effector 1B","#3F007D")) + volc_theme

# all days, c1b vs c3
volc_c1b_c3 = make_volcano_v2(agg_mat_c1b_c3, cluster_terms_c1b_c3, genes, lit_info_df = DEG_immgen_mem_eff_1000, label_color_pair_1 = c("Cluster 1b", "Memory Precursor B", "#08519C"), label_color_pair_2 = c("Cluster 3", "Effector 2B","#A50F15")) + volc_theme

# Save plots
ggsave(volc_c1a_c5_48h, filename = file.path(output_folder_new, "volc_c1a_c5_48h.pdf"), width=7, height=6)
ggsave(volc_c3_c4, filename = file.path(output_folder_new, "volc_c3_c4.pdf"), width=7, height=6)
ggsave(volc_c1b_c4, filename = file.path(output_folder_new, "volc_c1b_c4.pdf"), width=7, height=6)
ggsave(volc_c1b_c3, filename = file.path(output_folder_new, "volc_c1b_c3.pdf"), width=7, height=6)

```

# Make heatmap of DEGs using complex heatmap
For this, use agg_mat from all timepoints, all clusters DEG
```{r}
agg_mat <- agg_mat_all_clust_all_time
agg_mat_matrix <- as.matrix(agg_mat_all_clust_all_time)

# scale gene exprssion for each gene across all cells
agg_mat_matrix = t(apply(agg_mat_matrix, 1, function(x) {
    x = log10(x + .1)
    scale(x)
}))

# # TFs from gene module analysis
# genes_to_label_pos_3 <- which(rownames(agg_mat_matrix) %in% TF_Tcell )
# genes_to_label_3 <- rownames(agg_mat_matrix[genes_to_label_pos_3, ])

# # few genes for presentation figure
# g_fig <- c('GZMB','IL2RA','NFATC1','CCL4','NR4A3','NR4A2','MYC','IFNG','TBX21','ZEB2','IL12RB2','BHLHE40','BACH2','SLAMF6','LEF1','TCF7','SELL','MYB')

# more genes for paper figure
g_fig <- c('ID3', 'MYC', 'FOXO1', 'CCL3', 'CCL4', 'CXCR4', 'TOX', 'CXCR4', 'KLF6', 'GZMB','IL2RA','NFATC1','CCl3','CCL4','NR4A3','NR4A2','MYC','IFNG','TBX21','ZEB2','IL12RB2','BHLHE40','BACH2','SLAMF6','LEF1','TCF7','SELL','MYB', 'ZEB2', 'BCL2', 'PECAM1', 'BCL6', 'AFF3', 'EOMES')

genes_to_label_pos_7 <- which(rownames(agg_mat_matrix) %in% g_fig )
genes_to_label_7 <- rownames(agg_mat_matrix[genes_to_label_pos_7, ])
genes_to_label_7 <- str_to_title(tolower(genes_to_label_7)) # convert to capitalized lowercase

# name columns 
colnames(agg_mat_matrix)<-colnames(agg_mat)

ha = rowAnnotation( a7 = anno_mark(at= genes_to_label_pos_7, labels = genes_to_label_7, which="row", labels_gp = gpar(col= "black",fontsize = 12)))

set.seed(123)
split = data.frame(cutree(hclust(dist(agg_mat_matrix)), k = 4))
split = split %>% rename(clust = 1)
split$clust <- factor(split$clust, levels = c(3,4,2,1))

ht_list = Heatmap(agg_mat_matrix, 
                  col = colorRamp2(seq(from=-1, to=2, by = (2--1)/12)[1:12], viridis::viridis(12)), 
    name = "scaled_expr", 
    show_column_names = FALSE, 
    show_row_names = FALSE, 
    row_names_gp = gpar(fontsize = 3), 
    cluster_columns = FALSE, column_order = c('Cluster 2a', 'Cluster 5', 'Cluster 4', 'Cluster 2b', 'Cluster 3', 'Cluster 1a', 'Cluster 1b'),
    column_dend_reorder = FALSE, cluster_rows = TRUE, show_column_dend = TRUE, show_row_dend = TRUE,width = unit(8, "cm"),  
    clustering_method_rows = "ward.D2",
    row_gap = unit(1, "mm"),
    cluster_row_slices = FALSE, # manual ordering
    row_split = split,
    heatmap_legend_param = list(title = "Scaled log10(expr)"),
    right_annotation = ha,
    row_title = NULL
    # left_annotation = rowAnnotation(foo = anno_block(gp = gpar(fill = 2:4),
        # labels_gp = gpar(col = "white", fontsize = 10))), row_km=4, row_gap = unit(0, "mm"), cluster_row_slices = FALSE
    # row_order = unlist(new_order)
    
)



ht_list = draw(ht_list)

pdf(file.path(output_folder_new, "ht_Tcell_5k_aligned_7clust_400q_GOI-extra_ordered_20221020.pdf"), width=6, height=6)
draw(ht_list)
dev.off()

```

















# DEG between trajectories
Which genes vary between trajectories, after batching out pseudotime?
```{r}

compare_trajectory_DEG <- function(cds){
  
  # first remove genes expressed in < 50 cells  
  cds <- detect_genes(cds)
  cds_subset <- cds[which(rowData(cds)$num_cells_expressed > 50), ]
  
  # run fit models
  gene_fits  = fit_models(cds_subset, model_formula_str = "~trajectory + Tcell_branch_pseudotime_new")
  fit_coefs= coefficient_table(gene_fits)
  cluster_terms <- fit_coefs%>% filter(term != "(Intercept)")
  cluster_terms <-cluster_terms %>% filter (q_value < 0.05) %>%
    dplyr::select(gene_short_name, term, q_value, estimate)
  
  return(cluster_terms)
}


#load cds already clustered
cds<- cds_all

# add cluster and pseudotime info to cds
cds$tcell_cluster <- df_cell$Tcell_5k_aligned_7clust

df_cell[match(df_cell_E1$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E1$Tcell_branch_pseudotime_new
df_cell[match(df_cell_E2$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E2$Tcell_branch_pseudotime_new
df_cell[match(df_cell_MP$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_MP$Tcell_branch_pseudotime_new
cds$Tcell_branch_pseudotime_new <- df_cell$Tcell_branch_pseudotime_new

# remove the early activated cluster cells, which is common to all three trajectories
cds_subset <- cds[, colData(cds)$tcell_cluster != '2a' ]
# add trajectory ID column 
colData(cds_subset)$trajectory <- 'none'
cds_subset$trajectory[(colData(cds_subset)$sample %in% df_cell_E1$sample)] <- 'E1'
cds_subset$trajectory[(colData(cds_subset)$sample %in% df_cell_E2$sample)] <- 'E2'
cds_subset$trajectory[(colData(cds_subset)$sample %in% df_cell_MP$sample)] <- 'MP'

# factor order trajectories
colData(cds_subset)$trajectory <- factor(colData(cds_subset)$trajectory, levels = c('MP', 'E1', 'E2') )

# run function to get DEG between trajectories
trajectory_terms_compare_to_MP <- compare_trajectory_DEG(cds_subset)
trajectory_terms_compare_to_MP_with_pseudotime <- trajectory_terms_compare_to_MP

# select only the terms related to trajectory (not pseudotime)
trajectory_terms_compare_to_MP <- trajectory_terms_compare_to_MP %>% filter(term != 'Tcell_branch_pseudotime_new')

trajectory_terms <- trajectory_terms_compare_to_MP

```

```{r}
# save
# saveRDS(trajectory_terms_compare_to_MP, file = file.path(output_folder_new, "trajectory_terms_compare_to_MP.Rda"))


```


# Plot between trajectory DEG
Try plotting these DEG over pseudotime, similar approach to the within trajectory DEG plotting 

Function for plotting trajectory DEG between trajectories, where each trajectory is a vertical cluster 
```{r}
plot_trajectory_DEG <- function(cds_full, df_cell_E1, df_cell_E2, df_cell_MP, terms, gene_list1, gene_list2, n_clust){
  
  # subset cds to include only input DE genes
  terms_top_n <- terms 
  
  cds_subset <- cds[rowData(cds)$gene_short_name %in% terms_top_n$gene_short_name, ]
  
  #### custom pseudotime heatmap
  
  genes <- rownames(rowData(cds_subset))
  genes_short_names <- rowData(cds_subset)$gene_short_name
  
  # generate matrix separately for each trajectory and then bind them together
  cds_E1 <- cds[, colData(cds)$sample %in% df_cell_E1$sample]
  cds_E2 <- cds[, colData(cds)$sample %in% df_cell_E2$sample]
  cds_MP <- cds[, colData(cds)$sample %in% df_cell_MP$sample]
  
  # add pseudotime column
  cds_E1$Tcell_branch_pseudotime_new <- df_cell_E1$Tcell_branch_pseudotime_new
  cds_E2$Tcell_branch_pseudotime_new <- df_cell_E2$Tcell_branch_pseudotime_new
  cds_MP$Tcell_branch_pseudotime_new <- df_cell_MP$Tcell_branch_pseudotime_new
  
  # get normalized counts, order by pseudotime
  pt.matrix_E1 <- normalized_counts((cds_E1)[match(genes,rownames(rowData(cds_E1))),order(colData(cds_E1)$Tcell_branch_pseudotime_new)], norm_method = "size_only")
  pt.matrix_E2 <- normalized_counts((cds_E2)[match(genes,rownames(rowData(cds_E2))),order(colData(cds_E2)$Tcell_branch_pseudotime_new)], norm_method = "size_only")
  pt.matrix_MP <- normalized_counts((cds_MP)[match(genes,rownames(rowData(cds_MP))),order(colData(cds_MP)$Tcell_branch_pseudotime_new)], norm_method = "size_only")
  
  #  spline
  pt.matrix_E1 <- t(apply(pt.matrix_E1,1,function(x){smooth.spline(x,df=3)$y}))
  pt.matrix_E2 <- t(apply(pt.matrix_E2,1,function(x){smooth.spline(x,df=3)$y}))
  pt.matrix_MP <- t(apply(pt.matrix_MP,1,function(x){smooth.spline(x,df=3)$y}))

  # rownames
  rownames(pt.matrix_E1) <- str_to_title(tolower(genes_short_names));
  rownames(pt.matrix_E2) <- str_to_title(tolower(genes_short_names));
  rownames(pt.matrix_MP) <- str_to_title(tolower(genes_short_names));
  
  # combine into one matrix
  pt.matrix <- cbind(pt.matrix_E1, pt.matrix_E2, pt.matrix_MP)
  
  # normalize z score across all trajectories 
  pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))
  
  # label columns by trajectory ID
  labels <- c(rep("E1", dim(pt.matrix_E1)[2]), 
              rep("E2", dim(pt.matrix_E2)[2]),
              rep("MP", dim(pt.matrix_MP)[2]))
  
  
   # genes to label in heatmap
  genes_to_label_pos <- which(rownames(pt.matrix) %in% gene_list1 )
  genes_to_label <- rownames(pt.matrix[genes_to_label_pos, ])
  
  genes_to_label_pos2 <- which(rownames(pt.matrix) %in% gene_list2 )
  genes_to_label2 <- rownames(pt.matrix[genes_to_label_pos2, ])
  
  # Gene annotations
  ha = rowAnnotation(
                    a1 = anno_mark(at= genes_to_label_pos, labels = genes_to_label, which="row", labels_gp = gpar(col= "black",fontsize = 12)),
                    a2=anno_mark(at= genes_to_label_pos2, labels = genes_to_label2, which="row", labels_gp = gpar(col= "black",fontsize = 12)))
  
  # clust dendrogram split
  split = data.frame(cutree(hclust(dist(pt.matrix)), k = n_clust))
  split = split %>% dplyr::rename(clust = 1)
  split$clust <- factor(split$clust, levels = c(2,3,1))
  
  # Row annotation of clusters
    dend_ann = rowAnnotation(Trajectory = anno_block(gp = gpar(fill = c("#3F007D","#A50F15","#08519C"), col = 0)))
  

  #Ward.D2 Hierarchical Clustering
  hthc <- Heatmap(
    pt.matrix,
    name                         = "z-score",
    column_split = labels,
     # col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
    col                          = colorRamp2(seq(from=-2, to=2, by = (2--2)/12)[1:12], viridis::viridis(12)),
    show_row_names               = FALSE,
    show_column_names            = FALSE,
    row_names_gp                 = gpar(fontsize = 10),
    clustering_method_rows = "ward.D2",
    clustering_method_columns = "ward.D2",
    row_title_rot                = 0,
    cluster_rows                 = TRUE,
    cluster_row_slices           = FALSE,
    cluster_columns              = FALSE,
    row_title = NULL,
    
    row_split = split,
    
    row_gap = unit(1, "mm"),
    right_annotation = ha
    
    # optional cluster slide annotation
    # left_annotation = dend_ann
    )

  return(list(hthc, pt.matrix))
  
}


```

# plot labels continuous legend version
```{r}
plot_labels_continous <- function (labels, x, y, cell_size = 0.5, cell_alpha = .5, cols = col_vector, with_labels = T) {
  df_cell = data.frame(Cluster = labels, tsne_1 = x, tsne_2 = y)
  cluster_label = df_cell %>% dplyr::group_by(Cluster) %>% 
    dplyr::summarise(mean_1 = median(tsne_1), mean_2 = median(tsne_2))
  if (with_labels) {
    g1 = (ggplot() + geom_point(data = df_cell, aes(x = tsne_1, 
                                                    y = tsne_2, color = Cluster), size = cell_size, stroke = 0, 
                                shape = 16, alpha = cell_alpha) + ggrepel::geom_text_repel(aes(x = cluster_label$mean_1, 
                                                                                      y = cluster_label$mean_2, label = (cluster_label$Cluster))) + 
            xlab("") + ylab("") + 
            # xlim(0,2) + ylim(0,2)+
            theme_void() 
            # coord_fixed() +
            # guides(color = guide_legend(title = "",
            #                             override.aes = list(size = 4, alpha = cell_alpha))) + theme(legend.title = element_text(size = 6),
            #                                                                                legend.text = element_text(size = 6), legend.margin = margin(0,
            #                                                                                                                                             -10, 0, 10), legend.key.width = unit(0.15, "in"),
            #                                                                                legend.key.height = unit(0.15, "in"), legend.position = "left")
          )
            }
  else {
    g1 = (ggplot() + geom_point(data = df_cell, aes(x = tsne_1, 
                                                    y = tsne_2, color = Cluster), size = cell_size, stroke = 0, 
                                shape = 16, alpha = cell_alpha) + xlab("") + ylab("") + 
            # xlim(0,2) + ylim(0,2)+
            theme_void() +
            # coord_fixed() +
            # guides(color = guide_legend(title = "", override.aes = list(size = 4, 
            #                                                             alpha = 1))) + 
            theme(legend.title = element_text(size = 6), 
                                                                                             legend.text = element_text(size = 6), legend.margin = margin(0, 
                                                                                                                                                          -10, 0, 10), legend.key.width = unit(0.15, "in"), 
                                                                                             legend.key.height = unit(0.15, "in"), legend.position = "left"))
  }
  return(g1)
}

```



# First plot pseudotime for each trajectory
```{r}
# Plot pseudotime on each branch separately
df_cell_plot <- df_cell 
df_cell_plot[!(df_cell_plot$sample %in% df_cell_E1$sample), ]$Tcell_branch_pseudotime_new <- NA
df_cell_plot[(df_cell_plot$sample %in% df_cell_E1$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E1$Tcell_branch_pseudotime_new
ps_E1 <- plot_labels(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, cell_size=1, with_labels = F) +
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8, discrete = FALSE, guide = "colourbar") +
    theme(legend.title = element_blank(), legend.position = "none") 
ps_E1_legend <- plot_labels_continous(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, cell_size=1, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8, discrete = FALSE, guide = "colourbar")

df_cell_plot <- df_cell 
df_cell_plot[!(df_cell_plot$sample %in% df_cell_E2$sample), ]$Tcell_branch_pseudotime_new <- NA
df_cell_plot[(df_cell_plot$sample %in% df_cell_E2$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E2$Tcell_branch_pseudotime_new
ps_E2 <- plot_labels(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, cell_size=1, with_labels = F) +
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8) +
    theme(legend.title = element_blank(), legend.position = "none")
ps_E2_legend <- plot_labels_continous(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, cell_size=1, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8, discrete = FALSE, guide = "colourbar")

df_cell_plot <- df_cell 
df_cell_plot[!(df_cell_plot$sample %in% df_cell_MP$sample), ]$Tcell_branch_pseudotime_new <- NA
df_cell_plot[(df_cell_plot$sample %in% df_cell_MP$sample), ]$Tcell_branch_pseudotime_new <- df_cell_MP$Tcell_branch_pseudotime_new
ps_MP <- plot_labels(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, cell_size=1, with_labels = F) +
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8) +
  theme(legend.title = element_blank(), legend.position = "none") 
ps_MP_legend <- plot_labels_continous(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, cell_size=1, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8, discrete = FALSE, guide = "colourbar")







ps_E1
ps_E2
ps_MP
ps_E1_legend
ps_E2_legend
ps_MP_legend

# save_plot(ps_E1, filename = file.path(output_folder_new, "ps_E1_0c2a.pdf"), base_height = 3)
# save_plot(ps_E2, filename = file.path(output_folder_new, "ps_E2_0c2a.pdf"), base_height = 3)
# save_plot(ps_MP, filename = file.path(output_folder_new, "ps_MP_0c2a.pdf"), base_height = 3)
save_plot(ps_E1_legend, filename = file.path(output_folder_new, "ps_E1_0c2a_legend.pdf"), base_height = 3)

type(df_cell$Tcell_branch_pseudotime_new)

```

# Compare between trajectory DEG to TF modules
```{r}

# load 
trajectory_terms <- readRDS(file.path(output_folder_new, "trajectory_terms_compare_to_MP.Rda"))

GoI <- c(TF_Tcell, 'GZMB','IL2RA','NFATC1','CCl3','CCL4','NR4A3','NR4A2','MYC','IFNG','TBX21','ZEB2','IL12RB2','BHLHE40','BACH2','SLAMF6','LEF1','TCF7','SELL','MYB', 'ZEB2', 'BCL2', 'KLF6', 'PECAM1', 'BCL6', 'IL2RG', 'IL15RA', 'IL7RA','IFNGR1', 'IFNAR2')


# sort by q value and get distinct genes
trajectory_terms_distinct <- trajectory_terms %>% arrange(q_value) %>% distinct(gene_short_name, .keep_all = TRUE)
trajectory_terms_top_400 <- trajectory_terms_distinct %>% top_n(-400, q_value)
trajectory_terms_top_200 <- trajectory_terms_distinct %>% top_n(-200, q_value)
trajectory_terms_GoI <- trajectory_terms_distinct %>% filter(gene_short_name %in% GoI) %>% filter(q_value < 1e-10)
trajectory_terms_top_200_GoI <- rbind(trajectory_terms_top_200, trajectory_terms_GoI) %>% distinct(gene_short_name, .keep_all = TRUE) 


# # other ways of selecting top trajectory terms
# trajectory_terms_thresh <- trajectory_terms_distinct %>% filter(q_value < 1e-20)
# trajectory_terms_thresh$abs_estimate <- abs(trajectory_terms_thresh$estimate)
# trajectory_terms_thresh_top200est <- trajectory_terms_thresh %>% top_n(200, abs_estimate)

# which TFs from TF module analysis are in the between trajectory DEG?
trajectory_terms_TFmod <- trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net$TF ), ]
trajectory_terms_targets <- trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net$linked_gene), ]
trajectory_terms_TFmod_in_top200_GoI <- trajectory_terms_top_200_GoI[which(trajectory_terms_top_200_GoI$gene_short_name %in% df_TF_net$TF ), ]

# which linked genes are in the between trajectory DEG?
# trajectory_terms_targets <- trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net$linked_gene), ]
# trajectory_terms_targets_top100 <- trajectory_terms_targets %>% top_n(-100, q_value)
# df_TF_net_for_TFmod_in_top200 <- df_TF_net %>% filter(TF %in% trajectory_terms_TFmod_in_top200$gene_short_name)
# trajectory_terms_targets_for_TFmod_in_top200 <- trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net_for_TFmod_in_top200$linked_gene), ] %>% top_n(-100, q_value)

# which linked genes are in the between trajectory DEG and are targets of the key TFs
trajectory_terms_targets_top100 <- trajectory_terms_targets %>% top_n(-100, q_value)

# show top linked genes for each TF of interest
df_TF_net_TCF7 <- df_TF_net %>% filter(TF == 'TCF7')
df_TF_net_TBX21 <- df_TF_net %>% filter(TF == 'TBX21')
df_TF_net_BHLHE40 <- df_TF_net %>% filter(TF == 'BHLHE40')
df_TF_net_EGR1 <- df_TF_net %>% filter(TF == 'EGR1')
df_TF_net_EOMES <- df_TF_net %>% filter(TF == 'EOMES')
trajectory_terms_targets_for_TCF7_EOMES_TBX21_EGR1 <- rbind(
  trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net_TCF7$linked_gene), ] %>% top_n(-15, q_value), 
  trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net_TBX21$linked_gene), ] %>% top_n(-15, q_value),
  trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net_EOMES$linked_gene), ] %>% top_n(-15, q_value),
  trajectory_terms_distinct[which(trajectory_terms_distinct$gene_short_name %in% df_TF_net_EGR1$linked_gene), ] %>% top_n(-40, q_value))
  



# write top 200 trajectory terms to txt for import to python --> Dynamo
# write(trajectory_terms_top200q$gene_short_name, file = "/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202101_output/trajectory_terms_top200q.txt")
```

# Run visualiztion of between trajectory DEG

2020.10.20 add downsampling of trajectories 

```{r}

cds <- cds_all

# set pseudotime = 0 for all cluster 2a cells 
df_cell_E1[which(df_cell_E1$Tcell_5k_aligned_7clust == '2a'), ]$Tcell_branch_pseudotime_new <-0
df_cell_E2[which(df_cell_E2$Tcell_5k_aligned_7clust == '2a'), ]$Tcell_branch_pseudotime_new <-0
df_cell_MP[which(df_cell_MP$Tcell_5k_aligned_7clust == '2a'), ]$Tcell_branch_pseudotime_new <-0

# # downsample df_cell so that each trajectory has same cell number
# min_traj_cellno <- min(nrow(df_cell_E1), nrow(df_cell_E2), nrow(df_cell_MP))
# df_cell_E1_sub <- df_cell_E1[sample(nrow(df_cell_E1), min_traj_cellno), ]
# df_cell_E2_sub <- df_cell_E2[sample(nrow(df_cell_E2), min_traj_cellno), ]
# df_cell_MP_sub <- df_cell_MP[sample(nrow(df_cell_MP), min_traj_cellno), ]


gene_list1 = trajectory_terms_TFmod$gene_short_name
gene_list2 = trajectory_terms_targets_for_TCF7_EOMES_TBX21_EGR1$gene_short_name
gene_list2 = gene_list2[-which(gene_list2 %in% gene_list1)]
DEG_bt_traj_hm <- plot_trajectory_DEG(cds, df_cell_E1, df_cell_E2, df_cell_MP, trajectory_terms_top_200_GoI, str_to_title(tolower(gene_list1)), str_to_title(tolower(gene_list2)), n_clust=3)
# DEG_bt_traj_hm <- plot_trajectory_DEG(cds, df_cell_E1_sub, df_cell_E2_sub, df_cell_MP_sub, trajectory_terms_top_200_GoI, str_to_title(tolower(gene_list1)), str_to_title(tolower(gene_list2)), n_clust=3)


pdf(file.path(output_folder_new, "DEG_bt_traj_hm_0c2a_top200-GoI_add-eomes_TF-targets-only_20221020_v3_no_row_anno.pdf"), width=7, height=8)
draw(DEG_bt_traj_hm[[1]])
dev.off()

```


```{r}
gene_list1 <- c('GZMB', 'IL2RA', 'NR4A3', 'NR4A2', 'IFNG','TBX21', 'ZEB2', 'IL12RB2', 'LEF1', 'TCF7', 'SELL', 'MYB', 'EGR1', 'SLAMF6','FOXO1', 'BHLHE40')
out <- plot_trajectory_DEG(cds, df_cell_E1, df_cell_E2, df_cell_MP, trajectory_terms_top_200_GoI, gene_list1, gene_list2, n_clust=3)
hm <- out[[1]]
matrix <- out[[2]]



pdf(file.path(output_folder_new, "DEG_bt_traj_hm_0c2a_top200-GoI_label-SMALL_square.pdf"), width=6, height=6)
draw(hm)
dev.off()

```

# Get genes from each cluster in the heatmatp
```{r}
out <- plot_trajectory_DEG(cds, df_cell_E1, df_cell_E2, df_cell_MP, trajectory_terms_top_200_GoI, gene_list1, gene_list2, n_clust=4)
hm <- out[[1]]
matrix <- out[[2]]
pdf(file.path(output_folder_new, "DEG_bt_traj_hm_0c2a_top200-GoI_label-all_4clust.pdf"), width=6, height=10)
draw(DEG_bt_traj_hm_finer)
dev.off()

clusterlist = row_order(hm)
clu_df <- lapply(names(clusterlist), function(i){
  out <- data.frame(GeneID = rownames(matrix[clusterlist[[i]],]),
                                             Cluster = paste0("cluster", i),
                                             stringsAsFactors = FALSE)
     return(out)
   }) %>%  
     do.call(rbind, .)

# output each cluster list to txt file 
write(clu_df[which(clu_df$Cluster == 'cluster1'), ]$GeneID, file.path(output_folder_new, "hm_4clust_c1.txt"))
write(clu_df[which(clu_df$Cluster == 'cluster2'), ]$GeneID,file.path(output_folder_new, "hm_4clust_c2.txt"))
write(clu_df[which(clu_df$Cluster == 'cluster3'), ]$GeneID, file.path(output_folder_new, "hm_4clust_c3.txt"))
write(clu_df[which(clu_df$Cluster == 'cluster4'), ]$GeneID,file.path(output_folder_new, "hm_4clust_c4.txt"))

```

# Add gene heatmap cluster label to trajectory term df and save as csv for supplementary table
```{r}
gene_c1_MP <- unique(read.table(file.path(output_folder_new, "hm_4clust_c1.txt")))
gene_c2_E1_early <- unique(read.table(file.path(output_folder_new, "hm_4clust_c2.txt")))
gene_c3_E2 <- unique(read.table(file.path(output_folder_new, "hm_4clust_c3.txt")))
gene_c4_E1_late <- unique(read.table(file.path(output_folder_new, "hm_4clust_c4.txt")))

# add labels to trajectory terms 
trajectory_terms_top_200_GoI$heatmap_row_cluster <- 'none'
trajectory_terms_top_200_GoI[match(gene_c1_MP$V1, trajectory_terms_top_200_GoI$gene_short_name), ]$heatmap_row_cluster <- 'MP'
trajectory_terms_top_200_GoI[match(gene_c2_E1_early$V1, trajectory_terms_top_200_GoI$gene_short_name), ]$heatmap_row_cluster <- 'E1'
trajectory_terms_top_200_GoI[match(gene_c3_E2$V1, trajectory_terms_top_200_GoI$gene_short_name), ]$heatmap_row_cluster <- 'E2'
trajectory_terms_top_200_GoI[match(gene_c4_E1_late$V1, trajectory_terms_top_200_GoI$gene_short_name), ]$heatmap_row_cluster <- 'E1'

# save
write_csv(trajectory_terms_top_200_GoI, file.path(output_folder_new,'supplementary_tables', 'trajectory_terms_200n_GoI.csv'))
```

# Visualize expression of genes from heatmap cluster on UMAP
```{r}

cds <- cds_all

g <- read_lines("hm_4clust_c2.txt") # get gene list
g_ID <- rowData(cds[match(g, rowData(cds)$gene_short_name), ])$gene_id
g_mod <- cal_gene_module(cds, gene_list=g_ID, top_v = 3, low_v = -3)

# plot

df_cell <- df_cell %>%mutate(g_mod_expr = g_mod$Expression) 

# all timepoints
df_cell %>%
  ggplot() +
  geom_point(aes(x = umap_Tcell_5k_aligned_1, 
                 y = umap_Tcell_5k_aligned_2,
                 color = g_mod_expr)) +
          scale_alpha_continuous(range = c(0.5, 1), guide = "none") +
          theme_void() +
  viridis::scale_color_viridis(option = "viridis", 
                                         na.value = "grey80", 
                                         end = 0.8)

# day 2
df_cell %>%
  filter(RT_group %in% c('46-1', '46-2', '48-1', '48-2')) %>%
  ggplot() +
  geom_point(aes(x = umap_Tcell_5k_aligned_1, 
                 y = umap_Tcell_5k_aligned_2,
                 color = g_mod_expr)) +
          scale_alpha_continuous(range = c(0.5, 1), guide = "none") +
          theme_void() +
  viridis::scale_color_viridis(option = "viridis", 
                                         na.value = "grey80", 
                                         end = 0.8)

# day 4
df_cell %>%
  filter(RT_group %in% c('96-1', '96-2', '98-1', '98-2')) %>%
  ggplot() +
  geom_point(aes(x = umap_Tcell_5k_aligned_1, 
                 y = umap_Tcell_5k_aligned_2,
                 color = g_mod_expr)) +
          scale_alpha_continuous(range = c(0.5, 1), guide = "none") +
          theme_void() +
  viridis::scale_color_viridis(option = "viridis", 
                                         na.value = "grey80", 
                                         end = 0.8)

```



# Plot expression of gene activity over pseudotime - plot same gene for each branch in given plot
Take a representative selection of Effector, Memory, and cell cycle TFs and plot activity in pseudotime for each T cell activation branch

```{r}

# Compute the TF module expression (full expression)
unique_TF = unique(df_TF_net$TF)
# KA: for every TF, this function calculates the aggregated gene expression for all linked genes; output has rows = number of cells and cols = number of TFS 
TF_module_exprs_all = lapply(unique_TF, function(x) {
  gene_list = (df_TF_net %>% filter(TF == x))$linked_gene_id
  gene_expr = cal_gene_module(cds_all, gene_list, 3, -3)
  colnames(gene_expr) = c(x)
  return(gene_expr)
})

TF_module_exprs_all = do.call(cbind, TF_module_exprs_all)

# Compute the TF module expression (new expression)
TF_module_exprs_new = lapply(unique_TF, function(x) {
  gene_list = (df_TF_net %>% filter(TF == x))$linked_gene_id
  gene_expr = cal_gene_module(cds_new, gene_list, 3, -3)
  colnames(gene_expr) = c(x)
  return(gene_expr)
})

TF_module_exprs_new = do.call(cbind, TF_module_exprs_new)

```

# Plot expression of gene activity over pseudotime - plot same gene for each branch in given plot, but now without scaling the gene expression values 
Same as block above, but remove scaling from cal_gene_module
```{r}
# Compute the TF module expression (full expression)
unique_TF = unique(df_TF_net$TF)
# Compute the TF module expression (new expression)
TF_module_exprs_new_no_scale = lapply(unique_TF, function(x) {
  gene_list = (df_TF_net %>% filter(TF == x))$linked_gene_id
  gene_expr = cal_gene_module_no_scale(cds_new, gene_list, 3, -3)
  colnames(gene_expr) = c(x)
  return(gene_expr)
})

TF_module_exprs_new_no_scale = do.call(cbind, TF_module_exprs_new_no_scale)

```



```{r}
dim(TF_module_exprs_new)
dim(TF_module_exprs_all)
dim(cds_all)
dim(cds_new)
```


```{r}
all(cds_new$sample == rownames(cds_all$sample))
all(cds_new$sample == rownames(TF_module_exprs_new))
all(cds_all$sample == rownames(TF_module_exprs_all))
rownames(TF_module_exprs_new) = cds_new$sample
rownames(TF_module_exprs_all) = cds_all$sample
```


```{r}

branch_colors <- c("#3F007D","#A50F15","#08519C") # E1, E2, MPO
# ordered
color_cluster_new = c("cyan4", "#08519C",
                      "#969696", "#cc6541", "#A50F15",
                      "#3F007D", "#ab62c0")

 ggtheme<- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), plot.background = element_blank(), axis.line = element_line(colour = "black"), text=element_text(size=20), axis.text = element_text(size=18) )

# make sure that umap coordinates and TF experession values are in the same order
TF_module_exprs_new <- TF_module_exprs_new[match(df_cell$sample, rownames(TF_module_exprs_new)), ]

# new
plot_TF_in_pseudotime_across_branch <- function(df_cell_1, df_cell_2, df_cell_3, TF_module_exprs_df, gene_name_list, branch_names, branch_colors = branch_colors) {
  
  
  TF_module_exprs <- TF_module_exprs_df %>% dplyr::select(gene_name_list)
  TF_module_exprs$sample <- row.names(TF_module_exprs)
  TF_exprs_1 <- TF_module_exprs[match(df_cell_1$sample, TF_module_exprs$sample), ]
  df_cell_1 <- cbind(df_cell_1, TF_exprs_1[gene_name_list])
  df_cell_1$branch <- branch_names[1] 
  TF_exprs_2 <- TF_module_exprs[match(df_cell_2$sample, TF_module_exprs$sample), ]
  df_cell_2 <- cbind(df_cell_2, TF_exprs_2[gene_name_list])
  df_cell_2$branch <- branch_names[2] 
  TF_exprs_3 <- TF_module_exprs[match(df_cell_3$sample, TF_module_exprs$sample), ]
  df_cell_3 <- cbind(df_cell_3, TF_exprs_3[gene_name_list])
  df_cell_3$branch <- branch_names[3] 
  df_cell <- bind_rows(df_cell_1,df_cell_2, df_cell_3)
  df_cell <- melt(df_cell, id.vars = c("sample", "umap_Tcell_5k_aligned_1", "umap_Tcell_5k_aligned_2", "Tcell_branch_pseudotime_new", "branch"), measure.vars = gene_name_list)
  
  g1 <- ggplot(df_cell, aes(x=Tcell_branch_pseudotime_new, y=value, color = branch)) + 
    geom_smooth(se=F, method=loess) +
    # geom_smooth(se=F) +
    xlab('T cell activation pseudotime')+
    ylab('Scaled activity')+
    ggtheme + 
    scale_color_manual("", values = branch_colors) +
    ggtitle(gene_name_list)
  g1
}

# df_cell set U cluster = pseudotime 0
# set pseudotime = 0 for all cluster 2a cells 
df_cell_E1[which(df_cell_E1$Tcell_5k_aligned_7clust == '2a'), ]$Tcell_branch_pseudotime_new <-0
df_cell_E2[which(df_cell_E2$Tcell_5k_aligned_7clust == '2a'), ]$Tcell_branch_pseudotime_new <-0
df_cell_MP[which(df_cell_MP$Tcell_5k_aligned_7clust == '2a'), ]$Tcell_branch_pseudotime_new <-0

# Plot gene activity for multiple branches in pseudotime
g3 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('BHLHE40'), branch_names = c('E1','E2','MP'), branch_colors)
g5 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('TCF7'), branch_names = c('E1','E2','MP'), branch_colors)
g6 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('TBX21'), branch_names = c('E1','E2','MP'), branch_colors)
g7 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('EGR1'), branch_names = c('E1','E2','MP'), branch_colors)
g8 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('STAT3'), branch_names = c('E1','E2','MP'), branch_colors)
g9 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('RUNX2'), branch_names = c('E1','E2','MP'), branch_colors)
g10 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('EOMES'), branch_names = c('E1','E2','MP'), branch_colors)
g11 <- plot_TF_in_pseudotime_across_branch(df_cell_E1, df_cell_E2, df_cell_MP, TF_module_exprs_new, gene_name_list = c('MYB'), branch_names = c('E1','E2','MP'), branch_colors)

# save
# save_plot(g3, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_BHLHE40_0c2a.pdf"), base_height = 3)
# save_plot(g5, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_TCF7_0c2a.pdf"), base_height = 3)
# save_plot(g6, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_TBX21_0c2a.pdf"), base_height = 3)
# save_plot(g7, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_EGR1_0c2a.pdf"), base_height = 3)
# save_plot(g8, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_STAT3_0c2a.pdf"), base_height = 3)
# save_plot(g9, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_RUNX2_0c2a.pdf"), base_height = 3)
# save_plot(g10, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_EOMES_0c2a.pdf"), base_height = 3)
# save_plot(g11, filename = file.path(output_folder_new, "pseudotime_gene_plots/activity_MYB_0c2a.pdf"), base_height = 3)

# save as cowplot grid as alternative
TF_act_grid <-plot_grid(g6, g7, g10, g5, ncol = 1, nrow = 4, align = "h", base_height = 6)
TF_act_grid

pdf(file.path(output_folder_new, "pseudotime_gene_plots/TF_act_grid.pdf"))
TF_act_grid
dev.off()

```

# Plot expression of gene LEVELS over pseudotime - plot same gene for each branch in given plot
Take a representative selection of Effector, Memory, and cell cycle TFs and plot activity in pseudotime for each T cell activation branch 
```{r}



# make sure that umap coordinates and TF experession values are in the same order
TF_module_exprs_new <- TF_module_exprs_new[match(df_cell$sample, rownames(TF_module_exprs_new)), ]

# new
plot_gene_levels_in_pseudotime_across_branch <- function(cds, df_cell_1, df_cell_2, df_cell_3, gene, branch_names, branch_colors) {
  
  # get scaled expression of gene
  gene_expr = cal_gene_module(cds, rowData(cds_new[rowData(cds)$gene_short_name == gene, ])$gene_id, 3, -3)
  
  gene_expr$sample <- row.names(gene_expr)
  gene_expr_1 <- gene_expr[match(df_cell_1$sample, gene_expr$sample), ]
  df_cell_1 <- cbind(df_cell_1, gene_expr_1['Expression'])
  df_cell_1$branch <- branch_names[1] 
  gene_expr_2 <- gene_expr[match(df_cell_2$sample, gene_expr$sample), ]
  df_cell_2 <- cbind(df_cell_2, gene_expr_2['Expression'])
  df_cell_2$branch <- branch_names[2] 
  gene_expr_3 <- gene_expr[match(df_cell_3$sample, gene_expr$sample), ]
  df_cell_3 <- cbind(df_cell_3, gene_expr_3['Expression'])
  df_cell_3$branch <- branch_names[3] 
  df_cell <- bind_rows(df_cell_1,df_cell_2, df_cell_3)
  df_cell <- melt(df_cell, id.vars = c("sample", "umap_Tcell_5k_aligned_1", "umap_Tcell_5k_aligned_2", "Tcell_branch_pseudotime_new", "branch"), measure.vars = 'Expression')
  
  g1 <- ggplot(df_cell, aes(x=Tcell_branch_pseudotime_new, y=value, color = branch)) + 
    # geom_smooth(se=FALSE) +
    geom_smooth(se=F, method=loess) +
    xlab('T cell activation pseudotime')+
    ylab('Scaled expression')+
    theme(text = element_text(size = 20)) + 
    ggtitle(gene) +
    scale_color_manual("",values=branch_colors)
  
  g1
}

# Plot gene activity for multiple branches in pseudotime
# all
g3 <- plot_gene_levels_in_pseudotime_across_branch(cds_all, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'BHLHE40', branch_names = c('E1','E2','MP'), branch_colors)
g5 <- plot_gene_levels_in_pseudotime_across_branch(cds_all, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'TCF7', branch_names = c('E1','E2','MP'), branch_colors)
g17 <- plot_gene_levels_in_pseudotime_across_branch(cds_all, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'TBX21', branch_names = c('E1','E2','MP'), branch_colors)
g18 <- plot_gene_levels_in_pseudotime_across_branch(cds_all, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'IFNG', branch_names = c('E1','E2','MP'), branch_colors)

# new
g3n <- plot_gene_levels_in_pseudotime_across_branch(cds_new, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'BHLHE40', branch_names = c('E1','E2','MP'), branch_colors)
g5n <- plot_gene_levels_in_pseudotime_across_branch(cds_new, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'TCF7', branch_names = c('E1','E2','MP'), branch_colors)
g17n <- plot_gene_levels_in_pseudotime_across_branch(cds_new, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'TBX21', branch_names = c('E1','E2','MP'), branch_colors)
g18n <- plot_gene_levels_in_pseudotime_across_branch(cds_new, df_cell_E1, df_cell_E2, df_cell_MP, gene = 'IFNG', branch_names = c('E1','E2','MP'), branch_colors)


# save
# all
# save_plot(g3, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_BHLHE40_0c2a.pdf"), base_height = 3)
# save_plot(g5, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_TCF7_0c2a.pdf"), base_height = 3)
# save_plot(g17, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_TBX12_0c2a.pdf"), base_height = 3)
# save_plot(g18, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_IFNG_0c2a.pdf"), base_height = 3)

#new
save_plot(g3n, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_BHLHE40_0c2a_new.pdf"), base_height = 3)
save_plot(g5n, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_TCF7_0c2a_new.pdf"), base_height = 3)
save_plot(g17n, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_TBX12_0c2a_new.pdf"), base_height = 3)
save_plot(g18n, filename = file.path(output_folder_new, "pseudotime_gene_plots/levels_IFNG_0c2a_new.pdf"), base_height = 3)
```
