---
title: "Gene_module_analysis - T cell"
author: "Kathleen Abadie"
---

```{r}
suppressMessages(library(cowplot))
suppressMessages(library(tidyverse))
suppressMessages(library(monocle3))
suppressMessages(library(plotly))
library(princurve)
library(irlba)

suppressMessages(library(umap))
suppressMessages(library(ggplot2))

theme_set(theme_cowplot(font_size=7, line_size = 0.5))
options(repr.plot.width=3, repr.plot.height=3)
rm(list = ls())

```


# load data, set input and output folders
```{r}

# base folder -- user needs to fill in
base <- "/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1"

# location to store output 
output_folder <- file.path(base, "202101_output")
output_folder_new <- file.path("202108_output")

# input data location 
input_folder <- file.path(base, "scifate_dependencies")

# load scifate functions
source(file.path(input_folder, "2021_scifate_analysis_functions.R"))

# load monocle3 cds objects for full mRNA (cds_all) and new mRNA (cds_new)
cds_all <- readRDS(file.path(input_folder, "cds_all.RDS"))
cds_new <- readRDS(file.path(input_folder, "cds_new.RDS"))

# load the TF-gene linkage data identified
df_gene_TF_all_full <- readRDS(file.path(input_folder, "Link_full_mgi/df_gene_TF_all.RDS"))
df_gene_TF_all_new <- readRDS(file.path(input_folder, "Link_new_mgi/df_gene_TF_all.RDS"))

# combine links from full and new mRNA, removing duplicates
duprows <- which(!is.na(match(df_gene_TF_all_new$TF_link,df_gene_TF_all_full$TF_link)))
df_TF_net <- rbind(df_gene_TF_all_full,df_gene_TF_all_new[-duprows,] )

# T cell differentiation TFs identified from gene module analysis
TF_Tcell <- c('JUNB', 'MYC', 'TBX21', 'IRF4', 'EGR1', 'NFATC1', 'BMYC', 'MXI1', 'STAT5A', 'NFAT5', 'ELK3', 'EOMES', 'REL', 'BHLHE40', 'STAT3', 'RUNX2', 'FOXO3', 'MXD4', 'BCL11B', 'CUX1', 'GTF2I', 'FOXO1', 'FLI1', 'STAT1', 'CHD2', 'ZEB1', 'FOXN3', 'TCF7', 'LEF1', 'ELF1', 'MYB', 'IKZF1', 'TCF12')



```

#Plotting settings
```{r}
color_cluster_new = c("cyan4", "#08519C",
                      "#969696", "#cc6541", "#A50F15",
                      "#3F007D", "#ab62c0")
```

# Filter cds objects
```{r}
# filter cds by UMI
cds_all <- cds_all[,colData(cds_all)$mRNA_count > 5000] 
cds_new <- cds_new[,colData(cds_new)$mRNA_count > 5000] 

# df_cell 
df_cell <- data.frame("sample" = colnames(cds_all), "RT_group" = cds_all$RT_group)

# add newly_syn_UMIs col to cds_all 
colData(cds_all)$newly_syn_UMIs <- colData(cds_new)$newly_syn_UMIs

# make sure df_cell and cds_all and cds_new have same cells
cds_all <- cds_all[ , which(colData(cds_all)$sample %in% df_cell$sample)] 
cds_new <- cds_new[ , which(colData(cds_new)$sample %in% df_cell$sample)] 

# estimate size factors
cds_new = estimate_size_factors(cds_new)
cds_all = estimate_size_factors(cds_all)

```


# UMAP dimensional reduction and clustering using T cell differentiation module 
```{r}

# UMAP dimensionality reduction; preprocess using T cell differentiation module; align by cell cycle phase
cds_dr_gene_align <- function(cds, gene_module, reduction_method){
  cds <- preprocess_cds(cds, use_genes = as.character(gene_module))
  cds <- align_cds(cds, preprocess_method = 'PCA', alignment_group = 'Phase')   
  cds <- reduce_dimension(cds, reduction_method = reduction_method, umap.n_neighbors = 15L, umap.min_dist = 0.001) 
  cds <- cluster_cells(cds, reduction_method = reduction_method, random_seed = 2020)
  return(cds)
}

cds_dr_gene_align_0module <- function(cds, reduction_method){
  cds <- preprocess_cds(cds)
  # cds <- align_cds(cds, preprocess_method = 'PCA', alignment_group = 'Phase')   
  cds <- reduce_dimension(cds, reduction_method = reduction_method, umap.n_neighbors = 15L, umap.min_dist = 0.001) 
  cds <- cluster_cells(cds, reduction_method = reduction_method, random_seed = 2020)
  return(cds)
}

# function to add umap coordinates to df_cell
add_red_dim_coords <- function(cds, df_cell, umap1_name, umap2_name, reduction_method, multiplier){
  coords = reducedDim(cds, type = reduction_method)
  coords <- (coords - min(coords))/max(coords)
  df_cell[umap1_name] <- multiplier*coords[,1]
  df_cell[umap2_name] <- multiplier*coords[,2]
  return(df_cell) }

# T cell gene module
t_cell_linked_gene_names <- df_TF_net[which(df_TF_net$TF %in% TF_Tcell),]$linked_gene
t_cell_module <- rowData(cds_new[which(rowData(cds_new)$gene_short_name %in% c(TF_Tcell, t_cell_linked_gene_names)), ])$gene_id
length(unique(t_cell_module))

# run preprocess (without optional align), reduction, and clustering (full trancriptome) - without T cell TF module
cds_all_5k_0mod_0align <- cds_dr_gene_align_0module(cds_all, 'UMAP')

# run preprocess, optional align, reduction, and clustering (full trancriptome) - using T cell TF module
cds_all_5k_aligned <- cds_dr_gene_align(cds_all, unique(t_cell_module), 'UMAP')


# run preprocess, optional align, reduction, and clustering (new trancriptome) - without using T cell TF module
cds_new_5k_0mod_0align <- cds_dr_gene_align_0module(cds_new, 'UMAP')


# run preprocess, optional align, reduction, and clustering (new trancriptome) - using T cell TF module
cds_new_5k_aligned <- cds_dr_gene_align(cds_new, unique(t_cell_module), 'UMAP')

# or load previously saved cds_new_5k_aligned
cds_new_5k_aligned <- readRDS(file.path(input_folder, "cds_new_5k_aligned.RDS"))

# add umap coordinates to df_cell
df_cell <- add_red_dim_coords(cds_all_5k_0mod_0align, df_cell, 'umap_Tcell_all_5k_0mod_0align_1', 'umap_Tcell_all_5k_0mod_0align_2', 'UMAP', 1)
df_cell <- add_red_dim_coords(cds_all_5k_aligned, df_cell, 'umap_Tcell_all_5k_1', 'umap_Tcell_all_5k_2', 'UMAP', 1)
df_cell <- add_red_dim_coords(cds_new_5k_0mod_0align, df_cell, 'umap_Tcell_5k_0mod_0align_1', 'umap_Tcell_5k_0mod_0align_2', 'UMAP', 1)
df_cell <- add_red_dim_coords(cds_new_5k_aligned, df_cell, 'umap_Tcell_5k_aligned_1', 'umap_Tcell_5k_aligned_2', 'UMAP', 1)
```

# Plot UMAPs
```{r}

# plotting info for gene plots
GoI_gene_plot<- c('IFNG', 'GZMB','CCL3', 'TBX21', 'ZEB2','TOX','TCF7', 'SELL','SLAMF6')
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral"))) # color palette for ploting genes
 
# plot expression of genes of interest on aligned, new transcriptome
p_genes_new_aligned <- plot_cells(cds_new_5k_aligned, genes = toupper(GoI_gene_plot), show_trajectory_graph = FALSE, label_cell_groups=FALSE, cell_size=0.2) +
  scale_colour_gradientn(colours = myPalette(256))
p_genes_new_aligned

# plot umaps by day for different methods

# first randomize row order
set.seed(42)
rows <- sample(nrow(df_cell))
df_cell <- df_cell[rows, ]

# add day info to df_cell
df_cell <- df_cell %>% add_column(day = 0, .after = 'RT_group')
df_cell <- df_cell %>% mutate(
  day = case_when(startsWith(RT_group, "9") ~ "4",
                  startsWith(RT_group, "4") ~ "2",
                  startsWith(RT_group, "2") ~ "1"))

# add phase info to df_cell
df_cell$phase <- colData(cds_new_5k_aligned)$Phase

# plot
g_days_all_0mod <- plot_labels(df_cell$day, df_cell$umap_Tcell_all_5k_0mod_0align_1, df_cell$umap_Tcell_all_5k_0mod_0align_2, with_labels = F)
g_days_all <- plot_labels(df_cell$day, df_cell$umap_Tcell_all_5k_1, df_cell$umap_Tcell_all_5k_2, with_labels = F)
g_days_new_0mod <-plot_labels(df_cell$day, df_cell$umap_Tcell_5k_0mod_0align_1, df_cell$umap_Tcell_5k_0mod_0align_2, with_labels = F)
g_days_new <-plot_labels(df_cell$day, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, with_labels = F) 
g_days_new_phase <-plot_labels(df_cell$phase, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, with_labels = F) 

# save
save_plot(g_days_all_0mod, filename = file.path(output_folder_new, "g_days_all_0mod.pdf"), base_height = 3)
save_plot(g_days_all, filename = file.path(output_folder_new, "g_days_all.pdf"), base_height = 3)
save_plot(g_days_new_0mod, filename = file.path(output_folder_new, "g_days_new_0mod.pdf"), base_height = 3)
save_plot(g_days_new, filename = file.path(output_folder_new, "g_days_new.pdf"), base_height = 3)
save_plot(g_days_new_phase, filename = file.path(output_folder_new, "g_days_new_phase.pdf"), base_height = 3)
save_plot(p_genes_new_aligned, filename = file.path(output_folder, "p_genes_new_aligned.pdf"), base_height = 3)

```

# Visualize gene module expression
```{r}

# df_cell <- readRDS("/Users/kathleenabadie/Google Drive/1.Lab_starting_March_2018/5.Experiments_labwork/RNAseq/sci-FATE-seq/2019.06.11_scifate_expt1/202101_output/df_cell_5k_Tcell_7clust.RDS")

# effector module
TF_eff <- c('JUNB', 'MYC', 'TBX21', 'IRF4', 'EGR1', 'NFATC1', 'BMYC', 'MXI1', 'STAT5A', 'NFAT5', 'ELK3', 'EOMES', 'REL', 'BHLHE40', 'STAT3', 'RUNX2')
# mem module 
TF_mem <- c('FOXO3', 'MXD4', 'BCL11B', 'CUX1', 'GTF2I', 'FOXO1', 'FLI1', 'STAT1', 'CHD2', 'ZEB1', 'FOXN3', 'TCF7', 'LEF1', 'ELF1', 'MYB', 'IKZF1', 'TCF12')
TF_eff_module_sum <- rowSums(TF_module_exprs_new[, TF_eff])
TF_mem_module_sum <- rowSums(TF_module_exprs_new[, TF_mem])

# all, 0 mod
g_clust_all_0mod <- plot_labels(df_cell$Tcell_5k_aligned_7clust, df_cell$umap_Tcell_all_5k_0mod_0align_1, df_cell$umap_Tcell_all_5k_0mod_0align_2, with_labels = F) + 
  scale_color_manual("", values = color_cluster_new)
g_Eff_all_0mod <- plot_values(TF_eff_module_sum, df_cell$umap_Tcell_all_5k_0mod_0align_1, df_cell$umap_Tcell_all_5k_0mod_0align_2, "effector-like") 
g_MP_all_0mod <- plot_values(TF_mem_module_sum, df_cell$umap_Tcell_all_5k_0mod_0align_1, df_cell$umap_Tcell_all_5k_0mod_0align_2, "memory-like")

g_d2_all_0mod <- plot_labels(df_cell[startsWith(df_cell$RT_group, '4'), ]$Tcell_5k_aligned_7clust, df_cell[startsWith(df_cell$RT_group, '4'), ]$umap_Tcell_all_5k_0mod_0align_1, df_cell[startsWith(df_cell$RT_group, '4'), ]$umap_Tcell_all_5k_0mod_0align_2, with_labels = F, cell_size=2) +
  scale_color_manual("", values = color_cluster_new)
g_d4_all_0mod <- plot_labels(df_cell[startsWith(df_cell$RT_group, '9'), ]$Tcell_5k_aligned_7clust, df_cell[startsWith(df_cell$RT_group, '9'), ]$umap_Tcell_all_5k_0mod_0align_1, df_cell[startsWith(df_cell$RT_group, '9'), ]$umap_Tcell_all_5k_0mod_0align_2, with_labels = F, cell_size=2) +
  scale_color_manual("", values = color_cluster_new)


# all
g_clust_all <- plot_labels(df_cell$Tcell_5k_aligned_7clust, df_cell$umap_Tcell_all_5k_1, df_cell$umap_Tcell_all_5k_2, with_labels = F) + 
  scale_color_manual("", values = color_cluster_new)
g_Eff_all <- plot_values(TF_eff_module_sum, df_cell$umap_Tcell_all_5k_1, df_cell$umap_Tcell_all_5k_2, "effector-like") 
g_MP_all <- plot_values(TF_mem_module_sum, df_cell$umap_Tcell_all_5k_1, df_cell$umap_Tcell_all_5k_2, "memory-like")

g_d2_all <- plot_labels(df_cell[startsWith(df_cell$RT_group, '4'), ]$Tcell_5k_aligned_7clust, df_cell[startsWith(df_cell$RT_group, '4'), ]$umap_Tcell_all_5k_1, df_cell[startsWith(df_cell$RT_group, '4'), ]$umap_Tcell_all_5k_2, with_labels = F, cell_size=2) +
  scale_color_manual("", values = color_cluster_new)
g_d4_all <- plot_labels(df_cell[startsWith(df_cell$RT_group, '9'), ]$Tcell_5k_aligned_7clust, df_cell[startsWith(df_cell$RT_group, '9'), ]$umap_Tcell_all_5k_1, df_cell[startsWith(df_cell$RT_group, '9'), ]$umap_Tcell_all_5k_2, with_labels = F, cell_size=2) +
  scale_color_manual("", values = color_cluster_new)

# new, 0 mod
g_clust_new_0mod <- plot_labels(df_cell$Tcell_5k_aligned_7clust, df_cell$umap_Tcell_5k_0mod_0align_1, df_cell$umap_Tcell_5k_0mod_0align_2, with_labels = F) + 
  scale_color_manual("", values = color_cluster_new)
g_Eff_new_0mod <- plot_values(TF_eff_module_sum, df_cell$umap_Tcell_5k_0mod_0align_1, df_cell$umap_Tcell_5k_0mod_0align_2, "effector-like") 
g_MP_new_0mod <- plot_values(TF_mem_module_sum, df_cell$umap_Tcell_5k_0mod_0align_1, df_cell$umap_Tcell_5k_0mod_0align_2, "memory-like")

g_d2_new_0mod <- plot_labels(df_cell[startsWith(df_cell$RT_group, '4'), ]$Tcell_5k_aligned_7clust, df_cell[startsWith(df_cell$RT_group, '4'), ]$umap_Tcell_5k_0mod_0align_1, df_cell[startsWith(df_cell$RT_group, '4'), ]$umap_Tcell_5k_0mod_0align_2, with_labels = F, cell_size=2) +
  scale_color_manual("", values = color_cluster_new)
g_d4_new_0mod <- plot_labels(df_cell[startsWith(df_cell$RT_group, '9'), ]$Tcell_5k_aligned_7clust, df_cell[startsWith(df_cell$RT_group, '9'), ]$umap_Tcell_5k_0mod_0align_1, df_cell[startsWith(df_cell$RT_group, '9'), ]$umap_Tcell_5k_0mod_0align_2, with_labels = F, cell_size=2) +
  scale_color_manual("", values = color_cluster_new)

#new
g_clust_new <- plot_labels(df_cell$Tcell_5k_aligned_7clust, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, with_labels = F) + 
  scale_color_manual("", values = color_cluster_new)
g_Eff_new <- plot_values(TF_eff_module_sum, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, "effector-like") 
g_MP_new <- plot_values(TF_mem_module_sum, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, "memory-like")



# display
g_clust_all_0mod
g_Eff_all_0mod
g_MP_all_0mod
g_clust_all
g_Eff_all
g_MP_all
g_clust_new_0mod
g_Eff_new_0mod
g_MP_new_0mod
g_clust_new
g_Eff_new
g_MP_new

# save
# save_plot(g_clust_all_0mod, filename = file.path(output_folder, "g_clust_all_0mod.pdf"), base_height = 3)
# save_plot(g_Eff_all_0mod, filename = file.path(output_folder, "g_Eff_all_0mod.pdf"), base_height = 3)
# save_plot(g_MP_all_0mod, filename = file.path(output_folder, "g_Mp_all_0mod.pdf"), base_height = 3)
# save_plot(g_d2_all_0mod, filename = file.path(output_folder, "g_d2_all_0mod.pdf"), base_height = 3)
# save_plot(g_d4_all_0mod, filename = file.path(output_folder, "g_d4_all_0mod.pdf"), base_height = 3)
# 
# save_plot(g_clust_all, filename = file.path(output_folder, "g_clust_all.pdf"), base_height = 3)
# save_plot(g_Eff_all, filename = file.path(output_folder, "g_Eff_all.pdf"), base_height = 3)
# save_plot(g_MP_all, filename = file.path(output_folder, "g_Mp_all.pdf"), base_height = 3)
# save_plot(g_d2_all, filename = file.path(output_folder, "g_d2_all.pdf"), base_height = 3)
# save_plot(g_d4_all, filename = file.path(output_folder, "g_d4_all.pdf"), base_height = 3)
# 
# save_plot(g_clust_new_0mod, filename = file.path(output_folder, "g_clust_new_0mod.pdf"), base_height = 3)
# save_plot(g_Eff_new_0mod, filename = file.path(output_folder, "g_Eff_new_0mod.pdf"), base_height = 3)
# save_plot(g_MP_new_0mod, filename = file.path(output_folder, "g_Mp_new_0mod.pdf"), base_height = 3)
# save_plot(g_d2_new_0mod, filename = file.path(output_folder, "g_d2_new_0mod.pdf"), base_height = 3)
# save_plot(g_d4_new_0mod, filename = file.path(output_folder, "g_d4_new_0mod.pdf"), base_height = 3)
# 
# save_plot(g_clust_new, filename = file.path(output_folder, "g_clust_new.pdf"), base_height = 3)
# save_plot(g_Eff_new, filename = file.path(output_folder, "g_Eff_new.pdf"), base_height = 3)
# save_plot(g_MP_new, filename = file.path(output_folder, "g_Mp_new.pdf"), base_height = 3)

```











# Cluster T cell activation UMAP
```{r}
df_umap_Tcell = (df_cell %>% dplyr::select(umap_Tcell_5k_aligned_1, umap_Tcell_5k_aligned_2))
umap_mat = df_umap_Tcell
dataDist <- dist(umap_mat)
dataClust <- densityClust::densityClust(dataDist, gaussian = T)

g1 = (ggplot() 
      + geom_point(aes(x = dataClust$rho, y = dataClust$delta))
     )
g1

rho_thresh = 100 
delta_thresh = .5
g1 = (ggplot() 
      + geom_point(aes(x = dataClust$rho, y = dataClust$delta))
      + geom_vline(xintercept = rho_thresh)
      + geom_hline(yintercept = delta_thresh)
     )
g1

# find clusters
dataClust <- densityClust::findClusters(dataClust, 
                rho = rho_thresh, delta = delta_thresh)
df_umap = (df_umap_Tcell)
df_umap$Cluster = as.character(dataClust$clusters)
g1 = (plot_labels(df_umap$Cluster, df_umap$umap_Tcell_5k_aligned_1, df_umap$umap_Tcell_5k_aligned_2, with_labels = T) +
           scale_color_manual("", values = color_cluster_new)
     )
g1

# add cluster info to df_cell
df_cell$Tcell_5k_aligned_cluster <- df_umap$Cluster

```


# Subcluster memory cluster with kmeans
```{r}
# Cluster 1 (M)
set.seed(12)
df_subclust_1 <- df_cell %>% filter(Tcell_5k_aligned_cluster == 1)
df_umap <- df_subclust_1 %>% dplyr::select(umap_Tcell_5k_aligned_1, umap_Tcell_5k_aligned_2)
kmeans_res <- kmeans(df_umap, 2.5) 
df_umap$kmeans_clust <- as.factor(kmeans_res$cluster)
plot_labels(df_umap$kmeans_clust, df_umap$umap_Tcell_5k_aligned_1, df_umap$umap_Tcell_5k_aligned_2, with_labels = T)


# add to df_cell
df_subclust_1$Tcell_5k_aligned_cluster_kmeans = with(df_umap, ifelse(kmeans_clust == "1", "1a", 
                                          ifelse(kmeans_clust == "2", "1b", "unknown"))) 

# update df cell
df_cell$Tcell_5k_aligned_6clust <- df_cell$Tcell_5k_aligned
df_cell[which(df_cell$sample %in% df_subclust_1$sample), ]$Tcell_5k_aligned_6clust <- df_subclust_1$Tcell_5k_aligned_cluster_kmeans

# plot
g1 = (plot_labels(df_cell$Tcell_5k_aligned_6clust, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, with_labels = T) +
           scale_color_manual("", values = color_cluster_new)
     )
g1
```

# Subcluster early activation cluster with kmeans
```{r}
# Cluster 2 (Early)
set.seed(12)
df_subclust_2 <- df_cell %>% filter(Tcell_5k_aligned_cluster == 2)
df_umap <- df_subclust_2 %>% dplyr::select(umap_Tcell_5k_aligned_1, umap_Tcell_5k_aligned_2)
kmeans_res <- kmeans(df_umap, 2) 
df_umap$kmeans_clust <- as.factor(kmeans_res$cluster)
plot_labels(df_umap$kmeans_clust, df_umap$umap_Tcell_5k_aligned_1, df_umap$umap_Tcell_5k_aligned_2, with_labels = T)

# add to df_cell
df_subclust_2$Tcell_5k_aligned_cluster_kmeans = with(df_umap, ifelse(kmeans_clust == "1", "2b", 
                                          ifelse(kmeans_clust == "2", "2a", "unknown"))) 

# update df cell
df_cell$Tcell_5k_aligned_7clust <- df_cell$Tcell_5k_aligned_6clust
df_cell[which(df_cell$sample %in% df_subclust_2$sample), ]$Tcell_5k_aligned_7clust <- df_subclust_2$Tcell_5k_aligned_cluster_kmeans

# plot
g1 = (plot_labels(df_cell$Tcell_5k_aligned_7clust, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, with_labels = T) 
      +scale_color_manual("", values = color_cluster_new)
     )
g1
```

# Save df_cell containing cluster information
```{r}
# save df_cell
saveRDS(df_cell, file = file.path(output_folder, "df_cell_5k_Tcell_7clust.RDS"))
write_csv(df_cell, file = file.path(output_folder, "df_cell_5k_Tcell_7clust.csv"))

```


# Pseudotime for T cell activation module -- Run separately on each branch
```{r}
df_cell = readRDS(df_cell, file = file.path(output_folder, "df_cell_5k_Tcell_7clust.RDS"))

df_cell_E1 <- df_cell[which(df_cell$Tcell_5k_aligned_7clust %in% c('2a', '5', '4')), ]
df_cell_E2 <- df_cell[which(df_cell$Tcell_5k_aligned_7clust %in% c('2a', '2b', '3')), ]
df_cell_MP <- df_cell[which(df_cell$Tcell_5k_aligned_7clust %in% c('2a', '1a', '1b')), ]

# E1
df_umap_Tcell = as.matrix(df_cell_E1 %>% dplyr::select(umap_Tcell_5k_aligned_1, umap_Tcell_5k_aligned_2))
rownames(df_umap_Tcell) = df_cell_E1$sample
fit = principal_curve(df_umap_Tcell, maxit = 20)
Tcell_time = fit$lambda
df_cell_E1$Tcell_branch_pseudotime_new = 1- ( (max(Tcell_time) - Tcell_time) / max(Tcell_time) ) # added 1- to reverse pseudotime order

# E2
df_umap_Tcell = as.matrix(df_cell_E2 %>% dplyr::select(umap_Tcell_5k_aligned_1, umap_Tcell_5k_aligned_2))
rownames(df_umap_Tcell) = df_cell_E2$sample
fit = principal_curve(df_umap_Tcell, maxit = 20)
Tcell_time = fit$lambda
df_cell_E2$Tcell_branch_pseudotime_new = 1- ( (max(Tcell_time) - Tcell_time) / max(Tcell_time) ) # added 1- to reverse pseudotime order

# MP
df_umap_Tcell = as.matrix(df_cell_MP %>% dplyr::select(umap_Tcell_5k_aligned_1, umap_Tcell_5k_aligned_2))
rownames(df_umap_Tcell) = df_cell_MP$sample
fit = principal_curve(df_umap_Tcell, maxit = 20)
Tcell_time = fit$lambda
df_cell_MP$Tcell_branch_pseudotime_new = 1- ( (max(Tcell_time) - Tcell_time) / max(Tcell_time) ) # added 1- to reverse pseudotime order


# Add each branch pseudotime to main df_cell, for plotting 
df_cell$Tcell_branch_pseudotime_new <- 0
df_cell[match(df_cell_E1$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E1$Tcell_branch_pseudotime_new
df_cell[match(df_cell_E2$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E2$Tcell_branch_pseudotime_new
df_cell[match(df_cell_MP$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_MP$Tcell_branch_pseudotime_new


# Plot pseudotime on each branch separately
df_cell$Tcell_branch_pseudotime_new <- 0
df_cell[match(df_cell_E1$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E1$Tcell_branch_pseudotime_new
df_cell_plot <- df_cell 
df_cell_plot[!(df_cell_plot$sample %in% df_cell_E1$sample), ]$Tcell_branch_pseudotime_new <- NA
ps_E1 <- plot_labels(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8)

df_cell$Tcell_branch_pseudotime_new <- 0
df_cell[match(df_cell_E2$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_E2$Tcell_branch_pseudotime_new  
df_cell_plot <- df_cell 
df_cell_plot[!(df_cell_plot$sample %in% df_cell_E2$sample), ]$Tcell_branch_pseudotime_new <- NA
ps_E2 <- plot_labels(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8)

df_cell$Tcell_branch_pseudotime_new <- 0
df_cell[match(df_cell_MP$sample, df_cell$sample), ]$Tcell_branch_pseudotime_new <- df_cell_MP$Tcell_branch_pseudotime_new
df_cell_plot <- df_cell 
df_cell_plot[!(df_cell_plot$sample %in% df_cell_MP$sample), ]$Tcell_branch_pseudotime_new <- NA
ps_MP <- plot_labels(df_cell_plot$Tcell_branch_pseudotime_new, df_cell_plot$umap_Tcell_5k_aligned_1, df_cell_plot$umap_Tcell_5k_aligned_2, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8)

# Plot pseudotime of all branches together
ps_all <- plot_labels(df_cell$Tcell_branch_pseudotime_new, df_cell$umap_Tcell_5k_aligned_1, df_cell$umap_Tcell_5k_aligned_2, with_labels = F) + 
  viridis::scale_color_viridis(option = "inferno",
        name = "Values", na.value = "grey80", end = 0.8)

#save plots and RDS files 
save_plot(ps_E1, filename = file.path(output_folder, "ps_E1.pdf"), base_height = 3)
save_plot(ps_E2, filename = file.path(output_folder, "ps_E2.pdf"), base_height = 3)
save_plot(ps_MP, filename = file.path(output_folder, "ps_MP.pdf"), base_height = 3)
save_plot(ps_all, filename = file.path(output_folder, "ps_all.pdf"), base_height = 3)

saveRDS(df_cell, file = file.path(output_folder, "df_cell_5k_Tcell_7clust.RDS"))
saveRDS(df_cell_E1, file = file.path(output_folder, "df_cell_E1.RDS"))
saveRDS(df_cell_E2, file = file.path(output_folder, "df_cell_E2.RDS"))
saveRDS(df_cell_MP, file = file.path(output_folder, "df_cell_MP.RDS"))
```








